### ISSUE TITLE
Install empty BETYdb instances for testing + different projects

### ISSUE THREAD

**Issue by dlebauer** [terraref/computing-pipeline#24]
BETYdb is configured to be deployed and synced across many servers, and is used by distinct projects and research programs. 

The central database, betydb.org, is hosted at IGB (betydb.org; ebi-forecast.igb.illinois.edu). Other databases are listed on the ['Distributed BETYdb'](https://github.com/PecanProject/bety/wiki/Distributed-BETYdb) wiki page. The method used to synchronize and update individual servers is documented in [Database Synchronization is documented on the PEcAn wiki](https://github.com/PecanProject/pecan/wiki/Database-Synchronization).

For the TERRA program, we will create a network of BETYdb servers that specifically support Sorghum breeding. The TERRA Ref, TERRA-MEPP, and an instance for test data will be hosted on an open-stack production server (Nebula). TERRAref will be given the domain terraref.ncsa.illinois.edu/bety. This would allow us to launch a subset of databases focused on the specific needs of Sorghum, unlike the current instances that support terrestrial ecosystem modeling more broadly. The TERRA instances will retain interoperability with the PEcAn-BETYdb network, including the capacity to import and share data within or across networks.

The TERRA program is focused specifically on a large number of genotypes / lines within the genus _Sorghum_. There will be a small number of sites, and many high precision measurements of plots, subplot, and individual plant locations. The database will store data a more specific and restricted set of traits, data formats, and sites.

At deployment, all tables should be empty, and we can import specific records where necessary (e.g. the subset of Sorghum species that will be studied).

**Comment by dlebauer** [comment]
We can start with two instances of BETYdb: one for the TERRA REF team and the other for TERRA MEPP.

**Comment by dlebauer** [comment]
@robkooper could we start with PEcAn VMs for this? A third group has expressed interest, and I am realizing that this is a priority. They would need to be treated as production servers.

**Comment by dlebauer** [comment]
I've reserved a block of database host ids from 10-19 for TERRA-related projects (primarily for ease of keeping track).

(https://github.com/PecanProject/bety/wiki/Distributed-BETYdb)

|Institution | server | url | id | allocated primary key values| 
|---|---|---|---|---|
| TERRA test |  | | 10 | 10,000,000,001-11,000,000,000|
| TERRA Ref Danforth |  | | 11 | 11,000,000,001-12,000,000,000|
| TERRA MEPP UIUC |  | | 12 | 12,000,000,001-13,000,000,000|
| TERRA APS-GIEC Texas A&M |  | | 13 | 13,000,000,001-21,000,000,000|
| RESERVED |  | | 14-20 | 14,000,000,001-21,000,000,000|

**Comment by robkooper** [comment]
I will set them up differently from the pecan vm's. I want a lot more security focus on these servers than a standard vm.

**Comment by dlebauer** [comment]
@ankmichielsen can we ask the IGB CNRG group to host the database instance for TERRA-MEPP?

**Comment by ankmichielsen** [comment]
Not sure what you mean by hosting the database. Is this using biocluster?

**Comment by dlebauer** [comment]
I am not sure which server they will suggest - ebi-forecast would suffice but is nearing its end of life. They have a dedicated web server that may work. I can contact them, but wanted to check with you first.

**Comment by gsrohde** [comment]
For what it's worth, ebi-forecast is running CentOS 5, which is kind of oldish, a fact that complicated installing some things (Capybara-webkit and PostGIS-related packages for Ruby-on-Rails come to mind).  I'm assuming (hoping) that said dedicated server is running something more up-to-date.  Of course, a new Bety installation on ebi-forecast wouldn't require new installs of the supporting software.  But upgrading some of the software, should we wish to, might be difficult if not impossible.

**Comment by dlebauer** [comment]
Per discussion with Rob and David Slater, we've decided to start by installing an instance of BETYdb for data management in Steve Long's ARPA-E TERRA project on ebi-forecast (ultimately they plan to deploy OpenStack here, at which point we will migrate this).

David Slater will be the database / website administrator, so please help him get this up and running. 

* The MEPP database will be deployed to `ebi-forecast.igb.illinois.edu/bety-mepp`
* The host id for bety-mepp will be `8` (set `MYSITE=8` when running `load.bety.sh`) 
   * try `REMOTESITE=6` to install an [empty schema](http://file-server.igb.illinois.edu/~dlebauer/bety/bety.tar.gz). 
   * if `REMOTESITE=6` doesn't work, use `REMOTESITE=0` (default)
* set database name to `mepp_production`

Given that the server is already running an instance of BETYdb, hopefully David Slater will have enough information to (at least get started) deploying an instance of BETYdb for the MEPP project. Otherwise, I'll point him to this issue (#24) and our chatroom (https://gitter.im/PecanProject/bety) where he can ask questions.

As we deploy the TERRA network, we need to refine the documentation. David Slater - it would be great if you could take notes on the steps required for you to deploy this instance, and any other suggestions for simplifying our documentation. 

It appears that relevant, often redundant, and potentially inconsistent documentation can be found on multiple wiki pages, e.g. 

* https://github.com/PecanProject/bety/wiki/Installing-the-BetyDB-Rails-Application
* https://github.com/PecanProject/pecan/wiki/Installing-PEcAn#installing-bety
* https://github.com/PecanProject/bety/wiki/Distributed-BETYdb

(a copy of this comment will be sent to David Slater via email)

**Comment by dlebauer** [comment]
@gsrohde please also set up terra-ref (6) and terra-test (7) on pecandev. Please coordinate with David Slater on this so that he can efficiently deploy the MEPP instance described above.

**Comment by robkooper** [comment]
@dlebauer so no longer needed to set these up on openstack?

**Comment by dlebauer** [comment]
No, the goal would be to migrate to open stack when it is available. The
terra MEPP database will stay at IGB and migrate to their opens tack
instance once it is deployed.

If we can put production databases for terra ref on openstack now, that
would be great.

**Comment by gsrohde** [comment]
@dlebauer @robkooper 

I set up terra-ref on pecandev.  I have yet to do terra-test.

I wrote what I think is up-to-date documentation as I went along and put it on Gitbook at https://dlebauer.gitbooks.io/betydb-documentation/content/installing_a_production_copy_of_betydb.html.  (These instructions assume all of the supporting software is installed, as is the case on pecandev and ebi-forecast.)

To load the database schema, I used the db:structure:load rake task rather than running load.bety.sh (which script isn't part of the BETYdb repository).

**Comment by robkooper** [comment]
The most important step is missing which is what the load.bety.sh takes care of is to set the initial index for all newly genereted id's.

**Comment by gsrohde** [comment]
So if I use load.bety.sh to create the database but I want one with empty tables, do I have to manually truncate them, or is there a script option for creating a database copy with all the tables empty (except for schema_migrations, which needs to be fully populated)?  -r6 didn't work.

**Comment by robkooper** [comment]
I think you should be able to use load.bety.sh -f YES -m 6 after you created it

On Nov 25, 2015 2:53 PM, Scott Rohde <notifications@github.com> wrote:

So if I use load.bety.sh to create the database but I want one with empty tables, do I have to manually truncate them, or is there a script option for creating a database copy with all the tables empty (except for schema_migrations, which needs to be fully populated)? -r6 didn't work.

**Comment by gsrohde** [comment]
So you mean first run load.bety.sh with the -c option but without the -f and -m options and then re-run it as above?  I'm still seeing data in the tables.

**Comment by robkooper** [comment]
I just updated [load.bety.sh](https://github.com/PecanProject/pecan/blob/master/scripts/load.bety.sh) and added a new option to create (almost) empty database. You can use the following command to create an empty database: `./scripts/load.bety.sh -c YES -e YES -m 6 -r 0`.

This will copy only the following tables: formats machines mimetypes users

**Comment by gsrohde** [comment]
+schema_migrations.

This works well.

One additional request: can we have the guestuser user created (or copied) by default?

**Comment by gsrohde** [comment]
Is there any reason to copy the users table?  None of the data in the other copied tables refers to it, so I don't see the use of it.

Also, is "-m 6" just an example?  I assume (from reading https://github.com/PecanProject/bety/wiki/Distributed-BETYdb) that this should only be used for the copy that will live at terraref.ncsa.illinois.edu/bety and that other instances should use different numbers.  Should the copies that will live on pecandev use 99?

**Comment by robkooper** [comment]
users table is copied since all rows in all tables should have a created_by and updated_by row (I think). If they don't have it, I think it would be good to have it.

The -m 6 is an example and should be set to the appropriate host number, if not registered 99 (VM) is a good value (and is the default).

**Comment by robkooper** [comment]
Sure, just add -u YES to the options (see -h for help)

**Comment by gsrohde** [comment]
My point was that for publicly-available non-test deployments, we want the guestuser account to work, but we probably don't want an admin account carya/illinois as well.

**Comment by robkooper** [comment]
Use the brand new option -g

**Comment by gsrohde** [comment]
@robkooper @dlebauer 

RE the 4th and 5th comments preceding this one: I perhaps remember we added a created_by and/or updated_by column to one of the tables, but nearly all tables don't have them.  
So I still question the usefulness of copying the users table.  I'm not in a position to evaluate the usefulness of copying the formats, machines, and mimetypes tables, but I assume you have some good reasons for doing so.

**Comment by robkooper** [comment]
Because when we sync those databases with the current ones (which is what we want to be able to do) the constraints will kick in and prevent us from syncing if we duplicate the names of mimeformats etc.

**Comment by dslater82** [comment]
From https://dlebauer.gitbooks.io/betydb-documentation/content/installing_a_production_copy_of_betydb.html, I am not finding the createdb script anywhere.  Is it suppose to be in the /script folder?  Or is that step needed, according to https://github.com/PecanProject/bety/wiki/Installing-the-BetyDB-Rails-Application, you do not need to run createdb.  Does the update-betydb.sh create the database for you? I tried running and received the -e illegal option and it tries to create the bety database instead of bety-mepp.  

[root@ebi-forecast script]# ./update-betydb.sh -c YES -e YES -m 8 -r 6 -g -d bety-mepp
./update-betydb.sh: illegal option -- e
About to run ./load.bety.sh -c YES -e YES -m 8 -r 6 -g -d bety-mepp
psql: FATAL:  role "root" does not exist
Database bety does not exist, please create it:
(see https://github.com/PecanProject/pecan/wiki/Installing-PEcAn#installing-bety)
  sudo -u postgres createuser -d -l -P -R -S bety
  sudo -u postgres createdb -O bety bety

**Comment by dlebauer** [comment]
@dslater82 `createdb` is a psql command documented here: http://www.postgresql.org/docs/9.1/static/app-createdb.html

**Comment by gsrohde** [comment]
@dslater82 The -e option is new: It makes a nearly-empty database copy.  Try deleting load.bety.sh and then get a new copy by running ./update-bety.sh without options.  (I should really automate this somehow.)

**Comment by robkooper** [comment]
Also don't use 6, but use 0: `./load.bety.sh -c -e -m 8 -r 0 -g -d bety-mepp`

**Comment by dslater82** [comment]
I have the new site up at https://ebi-forecast.igb.illinois.edu/bety-mepp/.  A few things that should be added to the documentation.

mkdir /usr/local/bety-mepp/log
chown apache /usr/local/bety-mepp/log
chown apache /usr/local/bety-mepp/public/javascripts/cache
chown apache /usr/local/bety-mepp/tmp

The application wouldn't start without the cache folder permissions set correctly.  Having the log is nice.  I am assuming tmp is also needed.  I also had to run the load.bety.sh as the postgres user for it to work.  The gem file doesn't have i18n as a requirement.  I had to do gem install i18n manually.

**Comment by dslater82** [comment]
Also, the apache config is below.  This is different than what https://dlebauer.gitbooks.io/betydb-documentation/content/installing_a_production_copy_of_betydb.html has.

RackBaseURI /bety-mepp
        <Directory /usr/local/bety-mepp/public>
                Options FollowSymLinks
                AllowOverride None
                Order allow,deny
                Allow from all
        </Directory>
        <Location /bety-mepp>
                PassengerRuby /usr/local/ruby-2.1.5/bin/ruby
        </Location>

**Comment by gsrohde** [comment]
@dslater82 Regarding the need for making the log directory and chowning the log, cache, and tmp directories: The log directory would have been created automatically I think except that the apache user didn't have permission to do this.

I wrote the instructions under the assumption that the commands would be run as an ordinary user, not as root.  I mentioned this in footnote 1, but I guess I should move this out of the footnote and display it prominently at the start of the page.

Passenger will run as the user who is the owner of the startup file `config.ru` EXCEPT if the owner is root, then I guess it runs as apache.  So in the normal case, Passenger runs as the user who owns all the files and directories in the Rails tree (i.e., the user who ran `git clone`) and so automatically has the permissions it needs to do what it needs to do.

The Passenger documentation recommends creating a user account for each Rails app, but I guess we'll leave things as they are for now at least.

Regarding running load.bety.sh: Did you run this as root?  And if so, did you run `createdb` as root?

**Comment by dlebauer** [comment]
@robkooper where should the production instances of these go? On Nebula? Roger?

@gsrohde @max-zilla could you please work together to migrate these instances from the pecandev server?

**Comment by robkooper** [comment]
in process of creating an instance on nebula right now to host these.

**Comment by gsrohde** [comment]
@robkooper If you're taking this over, assign it to yourself.

**Comment by gsrohde** [comment]
Also, @dlebauer, do we need the instances that are on pecandev any more?  I'm assuming there is nothing we need to keep from them.

**Comment by robkooper** [comment]
not until it is copied and up and running on nebula

**Comment by gsrohde** [comment]
@max-zilla, @dlebauer, @robkooper I went about vastly simplifying the Apache servers.conf file, including the configuration pertaining to terra-test and terra-ref.  I think everything is working as it should, but let me know if you notice any problems.

**Comment by gsrohde** [comment]
David: FYI I updated the documentation, adding a step to make a dedicated OS user for the app.  This should prevent the permissions problems you saw.  But we can leave things as they are (that is, with Passenger running this instance as user “apache") for now.

I’m not sure why bundler didn’t install i18n.  Even though it is not in Gemfile, it _is_ listed in Gemfile.lock since it is required (indirectly) by the `rails` Gem, so installation should be automatic.

Also, it looks like you didn’t use the `—deployment` option to `bundle install`.  This gives each deployment instance its own set of Gems (under the vendor/ subdirectory), making the app instances more independent.  Since the other two BetyDB instances _do_ have their own bundles, this may not be a big deal unless and until we add another BETYdb instance on ebi-forecast.  And I’ve actually run into a problem only once, which was when I ran `bundle clean` on a BETYdb instance, which removes Gems not listed in Gemfile.lock.  At that time, our Redmine instances used the same version of Ruby as BETYdb, hence the same set of Gems.  Some of the Gems Redmine needed were removed.

**Comment by gsrohde** [comment]
David, a few notes:

RackBaseURI is deprecated in favor of PassengerBaseURI. (It still works though.)

While “Options FollowSymLinks” is necessary for the sole sym-link under `/usr/local/bety-mepp/public` (namely, `/usr/local/bety-mepp/public/images/lmodelout`), it is already enabled at the `<Directory “/“>` and `<Directory “/var/www/html”>` levels, so I don’t think you really need it here.

I’ve added the “Allow from all” requirement to the documentation.  (In fact, I’ve also revised the recommended deployment scheme in the documentation, using `Alias` directives instead of symlinking from the `/var/www/html` directory.  But again, I don’t see any compeling reason at this point to change existing deployments.  But should you need to add more …)

“Order allow,deny” probably isn’t needed.  More about this later.

**Comment by dlebauer** [comment]
Closing this issue - remaining tasks are in separate issues #56 and #24

* migrate terra-ref and terra-test to a stable open stack instance (nebula or clowder)  (#24)
* assign terraref.ncsa.illinois.edu domains to these (terra-ref, terra-sample, terra-dev) (#56)
